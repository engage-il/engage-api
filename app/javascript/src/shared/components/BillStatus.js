// @flow
import * as React from 'react'
import { createFragmentContainer, graphql } from 'react-relay'
import { last, range } from 'lodash'
import styled, { css } from 'react-emotion'
import { mixins, colors } from '@/styles'
import type { Bill, Step, StepActor } from '@/types'

const { floor } = Math

const SEGMENTS_LENGTH = 6
const SEGMENTS_LEAD = 0
const SEGMENTS_FOLLOW = 1

type Props = {
  bill: Bill
}

type Segment = {
  name: string,
  isLine: boolean,
  isActive: boolean
}

function hasPrimaryActor ({ actor }: Step) {
  switch (actor) {
    case 'LOWER':
    case 'UPPER':
    case 'GOVERNOR':
      return true
    default:
      return false
  }
}

function getLeadActor (steps: Step[]): ?StepActor {
  return steps.length > 0 ? steps[0].actor : null
}

function isOrderedByActor (prev: Step, next: Step) {
  return prev.actor !== next.actor && prev.actor !== 'GOVERNOR'
}

function chunkByActor (steps: Step[]): Step[][] {
  return steps.reduce((chunks, step) => {
    const chunk = last(chunks)
    const other = chunk && last(chunk)

    if (other == null || step.actor !== other.actor) {
      chunks.push([step])
    } else if (step.action !== other.action) {
      // we need to ignore duplicate actions for now, but this should
      // probably be done on the backend
      chunk.push(step)
    }

    return chunks
  }, [])
}

function findLastSequence (steps: Step[]) {
  const lead = getLeadActor(steps)
  if (lead == null) {
    return []
  }

  const actors = {}
  return chunkByActor(steps).reduceRight((sequence, chunk) => {
    const next = sequence[0]
    const prev = chunk[0]

    if (next == null || (!actors[lead] && !actors[prev.actor] && isOrderedByActor(prev, next))) {
      actors[prev.actor] = true
      return [...chunk, ...sequence]
    }

    return sequence
  }, [])
}

function getSegmentName (actor: StepActor) {
  switch (actor) {
    case 'LOWER':
      return 'House'
    case 'UPPER':
      return 'Senate'
    default: // 'GOVERNOR'
      return 'Governor'
  }
}

function initSegment (step: Step): Segment {
  return {
    name: getSegmentName(step.actor),
    isLine: step.action === 'RESOLVED',
    isActive: true
  }
}

function stubActor (lead: StepActor, index: number) {
  const slice = floor(index / 2)

  switch (slice) {
    case SEGMENTS_LEAD:
      return lead
    case SEGMENTS_FOLLOW:
      return lead === 'LOWER' ? 'UPPER' : 'LOWER'
    default:
      return 'GOVERNOR'
  }
}

function stubSegment (lead: StepActor, index: number) {
  return {
    name: getSegmentName(stubActor(lead, index)),
    isLine: index % 2 === 1,
    isActive: false
  }
}

function segmentsFromSteps (steps: Step[]): Segment[] {
  const lead = getLeadActor(steps) || 'LOWER'

  return range(0, SEGMENTS_LENGTH).map((i) => {
    const step = steps[i]
    if (step != null) {
      return initSegment(step)
    } else {
      return stubSegment(lead, i)
    }
  })
}

function segmentsFromBill ({ steps }: Bill) {
  return segmentsFromSteps(findLastSequence(steps.filter(hasPrimaryActor)))
}

function getKey ({ id }: Bill, index: number) {
  return `${id}-status-segment-${index}`
}

let BillStatus = function BillStatus ({ bill }: Props) {
  return (
    <Status>
      <Line isActive />
      {segmentsFromBill(bill).map(({ name, isLine, ...props }, i) => (
        isLine
          ? <Line key={getKey(bill, i)} {...props} />
          : <Name key={getKey(bill, i)} {...props}>{name}</Name>
      ))}
    </Status>
  )
}

const Status = styled.div`
  ${mixins.flexRow};

  align-items: center;
  align-self: stretch;
  max-width: 830px;
  height: 13px;
  margin-bottom: 15px;
`

const cap = css`
  width: 10px;
  height: 10px;
  background-color: ${colors.gray6};
  border-radius: 50%;
`

const active = css`
  background-color: ${colors.secondary};

  &:first-child, &:last-child {
    &:before {
      background-color: ${colors.secondary};
    }
  }
`

const Line = styled.div`
  ${mixins.flexRow};

  flex: 2;
  align-items: center;
  height: 2px;
  background-color: ${colors.gray6};

  &:last-child {
    justify-content: flex-end;
  }

  &:first-child, &:last-child {
    flex: 1;

    &:before {
      content: '';
      ${cap};
    }
  }

  ${({ isActive }) => isActive && active}
`

const Name = styled.span`
  margin: 0 10px;
  color: ${({ isActive }) => isActive ? colors.primary : colors.gray5};
  font-size: 12px;
  text-transform: uppercase;
`

BillStatus = createFragmentContainer(BillStatus, graphql`
  fragment BillStatus_bill on Bill {
    id
    steps {
      actor
      action
    }
  }
`)

export { BillStatus }
