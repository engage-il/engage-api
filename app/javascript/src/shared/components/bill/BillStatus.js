// @flow
import * as React from 'react'
import { createFragmentContainer, graphql } from 'react-relay'
import styled, { css } from 'react-emotion'
import { segmentsFromBill } from '@/functions'
import { mixins, colors } from '@/styles'
import type { Bill } from '@/types'

type Props = {
  bill: Bill
}

function getKey ({ id }: Bill, index: number) {
  return `${id}-status-segment-${index}`
}

let BillStatus = function BillStatus ({ bill }: Props) {
  return (
    <Status>
      <Line isActive />
      {segmentsFromBill(bill).map(({ name, isLine, ...props }, i) => (
        isLine
          ? <Line key={getKey(bill, i)} {...props} />
          : <Name key={getKey(bill, i)} {...props}>{name}</Name>
      ))}
    </Status>
  )
}

const Status = styled.div`
  ${mixins.flexRow};

  align-items: center;
  align-self: stretch;
  max-width: 830px;
  height: 13px;
  margin-bottom: 15px;
`

const cap = css`
  width: 10px;
  height: 10px;
  background-color: ${colors.gray6};
  border-radius: 50%;
`

const active = css`
  background-color: ${colors.secondary};

  &:first-child, &:last-child {
    &:before {
      background-color: ${colors.secondary};
    }
  }
`

const Line = styled.div`
  ${mixins.flexRow};

  flex: 2;
  align-items: center;
  height: 2px;
  background-color: ${colors.gray6};

  &:last-child {
    justify-content: flex-end;
  }

  &:first-child, &:last-child {
    flex: 1;

    &:before {
      content: '';
      ${cap};
    }
  }

  ${({ isActive }) => isActive && active}
`

const Name = styled.span`
  margin: 0 10px;
  color: ${({ isActive }) => isActive ? colors.primary : colors.gray5};
  font-size: 12px;
  text-transform: uppercase;
`

BillStatus = createFragmentContainer(BillStatus, graphql`
  fragment BillStatus_bill on Bill {
    id
    steps {
      actor
      action
    }
  }
`)

export { BillStatus }
