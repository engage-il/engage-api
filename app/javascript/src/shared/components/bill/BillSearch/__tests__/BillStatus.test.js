/* eslint-env jest */
import * as React from 'react'
import { shallow } from 'enzyme'
import { defaultsDeep } from 'lodash'
import { BillStatus } from '../BillStatus'

// subject
let subject

const defaults = {
  bill: {
    id: 'test-id',
    steps: [
      { actor: 'LOWER', action: 'INTRODUCED' },
      { actor: 'LOWER_COMMITTEE', action: 'INTRODUCED' },
      { actor: 'LOWER_COMMITTEE', action: 'RESOLVED' },
      { actor: 'LOWER', action: 'RESOLVED' },
      { actor: 'LOWER', action: 'RESOLVED' },
      { actor: 'UPPER', action: 'INTRODUCED' },
      { actor: 'UPPER_COMMITTEE', action: 'INTRODUCED' }
    ]
  }
}

function loadSubject (props = {}) {
  subject = shallow(<BillStatus {...defaultsDeep(props, defaults)} />)
}

// spec
afterEach(() => {
  subject = null
})

describe('#render', () => {
  it('shows the status', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })

  it('removes obsolete steps', () => {
    loadSubject({
      bill: {
        steps: [
          ...defaults.bill.steps,
          { actor: 'UPPER_COMMITTEE', action: 'RESOLVED' },
          { actor: 'UPPER', action: 'RESOLVED' },
          { actor: 'LOWER', action: 'INTRODUCED' }
        ]
      }
    })

    expect(subject).toMatchSnapshot()
  })
})
