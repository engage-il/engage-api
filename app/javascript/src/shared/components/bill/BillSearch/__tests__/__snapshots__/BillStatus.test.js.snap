// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render removes obsolete steps 1`] = `
<Styled(div)>
  <Styled(div)
    isActive={true}
  />
  <Styled(span)
    isActive={true}
    key="test-id-status-segment-0"
  >
    House
  </Styled(span)>
  <Styled(div)
    isActive={false}
    key="test-id-status-segment-1"
  />
  <Styled(span)
    isActive={false}
    key="test-id-status-segment-2"
  >
    Senate
  </Styled(span)>
  <Styled(div)
    isActive={false}
    key="test-id-status-segment-3"
  />
  <Styled(span)
    isActive={false}
    key="test-id-status-segment-4"
  >
    Governor
  </Styled(span)>
  <Styled(div)
    isActive={false}
    key="test-id-status-segment-5"
  />
</Styled(div)>
`;

exports[`#render shows the status 1`] = `
<Styled(div)>
  <Styled(div)
    isActive={true}
  />
  <Styled(span)
    isActive={true}
    key="test-id-status-segment-0"
  >
    House
  </Styled(span)>
  <Styled(div)
    isActive={true}
    key="test-id-status-segment-1"
  />
  <Styled(span)
    isActive={true}
    key="test-id-status-segment-2"
  >
    Senate
  </Styled(span)>
  <Styled(div)
    isActive={false}
    key="test-id-status-segment-3"
  />
  <Styled(span)
    isActive={false}
    key="test-id-status-segment-4"
  >
    Governor
  </Styled(span)>
  <Styled(div)
    isActive={false}
    key="test-id-status-segment-5"
  />
</Styled(div)>
`;
